### PR TITLE
bugfix for calculate_profile_data

### LINKED ISSUE #32: Bug in calculate_profile_data
`calculate_profile_data` - calculates a one dimensional diffraction profile for a structure.

**Describe the bug**
It does not return the correct multiplicity of generated planes

Problem is in line 237: `peaks[g_hkl] = [i_hkl, [tuple(hkl)], d_hkl]`

Planes  from the same family have the same spacing in reciprocal space (g_hkl) which means that planes from the same family overwrite each other.

**Possible solution** 
Instead of defining a dictionary of peaks, create list:
`peaks.append([g_hkl, i_hkl, [tuple(hkl)], d_hkl])` and then feed `peaks[2]` into get_unique_families. Currently working on that. 

**Notes**
Have tested `get_unique_families` separately and that function seems to work fine.

### PR DESCRIPTION
**Pull request by isabelwood100**
Changed the key of the dictionary to unique tuple(hkl) such that it no longer overwrites old entries. This will resolve #32 

Once the new plot is obtained a new bug is observed in get_plot of ProfileSimulation such that it plots all the Miller indices labels for all planes in that family on top of themselves at each Bragg peak rather than plotting the family so i'll look at fixing that.

### CODE REVIEW

**Comment by dnjohnstone** [comment]
Let us know when this is ready for a review @isabelwood100 ? Do you need input now, or you still know where you're going?

One thing is that there is a merge conflict - this will be resolved by merging this branch with the new master copy. Do you know how to do that?

**Comment by isabelwood100** [comment]
> Let us know when this is ready for a review @isabelwood100 ? Do you need input now, or you still know where you're going?
> 
> One thing is that there is a merge conflict - this will be resolved by merging this branch with the new master copy. Do you know how to do that?

In terms of #32 I believe this update has fixed it. Should I update the diffraction_simulation issue (multiple label plotting) on this branch too or make a new one since it is a different file?
I'm not sure about how to fix the merge conflict but i'll look into it now.

**Comment by dnjohnstone** [comment]
> In terms of #32 I believe this update has fixed it. Should I update the diffraction_simulation issue (multiple label plotting) on this branch too or make a new one since it is a different file?
> I'm not sure about how to fix the merge conflict but i'll look into it now.

Great - please fix the multiple label plotting in this PR also. It's sufficiently similar in this case because it relates to the same general functionality and it was a bug that was hidden by the bug that you've fixed here (so kind of the same bug). The total amount of code change is also quite small so doing this only makes reviewing this PR easier rather than harder :)

In terms of the merge conflict, you need to:

1. update your local master version of the code to be level with pyxem/master and push that to isabelwood100/master
2. switch back to this branch and merge the master into it
3. push this branch back to isabelwood100/Izzy_diffsims_diffgen

Does that make sense?

**Comment by isabelwood100** [comment]
> > In terms of #32 I believe this update has fixed it. Should I update the diffraction_simulation issue (multiple label plotting) on this branch too or make a new one since it is a different file?
> > I'm not sure about how to fix the merge conflict but i'll look into it now.
> 
> Great - please fix the multiple label plotting in this PR also. It's sufficiently similar in this case because it relates to the same general functionality and it was a bug that was hidden by the bug that you've fixed here (so kind of the same bug). The total amount of code change is also quite small so doing this only makes reviewing this PR easier rather than harder :)
> 
> In terms of the merge conflict, you need to:
> 
> 1. update your local master version of the code to be level with pyxem/master and push that to isabelwood100/master
> 2. switch back to this branch and merge the master into it
> 3. push this branch back to isabelwood100/Izzy_diffsims_diffgen
> 
> Does that make sense?

Ive done 1. and 2., but i'm not sure what you mean by pushing this branch back to isabelwood100/Izzy_diffsims_diffgen as isn't this already that branch? Thanks.

**Comment by dnjohnstone** [comment]
It's ok, you've done it - push just meant send it from your local computer to the online version.

Your changes have however broken a test,but I think your code is right so you'll need to update the test

**Comment by isabelwood100** [comment]
> It's ok, you've done it - push just meant send it from your local computer to the online version.
> 
> Your changes have however broken a test,but I think your code is right so you'll need to update the test

I can see its failing the three checks Travis Cl, cont. intergration and coverage, but im not sure how to update those tests? I ran it with test_diffraction_generator and it seems to be passing that.

**Comment by dnjohnstone** [comment]
> I can see its failing the three checks Travis Cl, cont. intergration and coverage, but im not sure how to update those tests? I ran it with test_diffraction_generator and it seems to be passing that.

Ok - this might be one for an "in person" meeting because I'm not quite sure how you're running your tests, are you using pytest in the terminal? But if you click on details on Travis or continuous-integration you'll eventually get here: https://travis-ci.com/github/pyxem/diffsims/jobs/366323890

That shows you that "test_plot_profile_simulation" is failing, probably because of how you've updated get_plot() ?

File: diffsims/sims/diffraction_simulation.py [hunk 1]
```diff
@@ -200,10 +203,13 @@ def get_plot(self, g_max, annotate_peaks=True, with_labels=True, fontsize=12):
         fontsize : integer
             Fontsize for peak labels.
         """
+
+        label_hkl = get_unique_families(hkls.keys())
```
Comment (pc494) [inline diffsims/sims/diffraction_simulation.py]: at a glance, I think changing `hkls.keys()` to `self.hkls.keys()` might get you going? `hkls` is an attribute of a `ProfileSimulation`, which in this layout is "`self`"

Comment (dnjohnstone) [reply diffsims/sims/diffraction_simulation.py]: Now it looks like it should probably just be `self.hkls` because that's a list and not a dictionary?

**Comment by isabelwood100** [comment]
> > I can see its failing the three checks Travis Cl, cont. intergration and coverage, but im not sure how to update those tests? I ran it with test_diffraction_generator and it seems to be passing that.
> 
> Ok - this might be one for an "in person" meeting because I'm not quite sure how you're running your tests, are you using pytest in the terminal? But if you click on details on Travis or continuous-integration you'll eventually get here: https://travis-ci.com/github/pyxem/diffsims/jobs/366323890
> 
> That shows you that "test_plot_profile_simulation" is failing, probably because of how you've updated get_plot() ?

Yes im running pytest in the terminal - it was my mistake i was running the test for diffraction_generator not diffraction_simulation, so i can now see where its erroring.

**Comment by isabelwood100** [comment]
In the latest commit i've Included multiplicities in get_kinematical_intensities
Updated variable names in diffraction_generator to be consistent with sim_utils
Used get_kinematical_intensities in calculate_profile_data for consistency

But there are points where I have an idea of what needs to be done but don't have the knowledge of how to do it effectively so might need a pointer..

1) the unique hkls and m's need to be extracted from the unique_hkls_dict to form arrays which can be fed into get_unique_families, but in a way that the m's match up with the corresponding unique hkl

2) the g_hkls, excitation_error and g_indices have already been calculated in the previous method for all the hkls inb the sphere - do I need to create a new array including only the ones corresponding to the unique hkls, or when get_kinematical_intestites calls them will it automatically call the correct ones calculated for that given hkl?

Once those are added the intensities should be calculated correctly, and then the plotting section needs to be updated to get rid of get_unique_families in there as I don't think its needed there anymore.

**Comment by dnjohnstone** [comment]
@isabelwood100 is this ready for review? Are tests on their way?

**Comment by dnjohnstone** [comment]
Also @isabelwood100 do you know how to resolve the merge conflict?

**Comment by isabelwood100** [comment]
> Also @isabelwood100 do you know how to resolve the merge conflict?

I just had a look and i'm a bit stuck.

**Comment by isabelwood100** [comment]
> @isabelwood100 is this ready for review? Are tests on their way?

Yes and yes hopefully i'll have them by tomorrow.

**Comment by pc494** [comment]
> > Also @isabelwood100 do you know how to resolve the merge conflict?
> 
> I just had a look and i'm a bit stuck.

This branch doesn't "know" about #101, so github isn't sure if `sim_utils` should contain your new function here or the functions you moved there in your previous PR. Given that we want both it's sufficient to just delete out the lines added by github, which look like:
```
<<<<< name_of_branch
=====
===== master
```

**Comment by isabelwood100** [comment]
> > > Also @isabelwood100 do you know how to resolve the merge conflict?
> > 
> > 
> > I just had a look and i'm a bit stuck.
> 
> This branch doesn't "know" about #101, so github isn't sure if `sim_utils` should contain your new function here or the functions you moved there in your previous PR. Given that we want both it's sufficient to just delete out the lines added by github, which look like:
> 
> ```
> <<<<< name_of_branch
> =====
> ===== master
> ```

Great thank you!

**Comment by dnjohnstone** [comment]
Ok there are quite a lot of testing issues here @isabelwood100 - do you understand them and know how to work to fix them?

File: diffsims/tests/test_utils/test_sim_utils.py [hunk 2]
```diff
@@ -146,165 +140,32 @@ def test_uvtw_to_uvw(uvtw, uvw):
...
-        acceleration_voltage = 300000
-        lattice_size = np.array([2e-10, 3e-10])
-        miller_index = (1, 0, 0)
-        scattering_angle = diffraction_scattering_angle(
-            acceleration_voltage, lattice_size, miller_index
-        )
-        assert len(scattering_angle) == 2
-        sa_known = np.array([9.84e-3, 6.56e-3])
-        assert np.allclose(sa_known, scattering_angle, rtol=0.001)
+def test_get_intesnities_params():
```
Comment (pc494) [inline diffsims/tests/test_utils/test_sim_utils.py]: there is a repeated typo, should be "intensities"

File: diffsims/tests/test_utils/test_sim_utils.py [hunk 3]
```diff
@@ -146,165 +140,32 @@ def test_uvtw_to_uvw(uvtw, uvw):
...
+    struct = make_structure()
```
Comment (pc494) [inline diffsims/tests/test_utils/test_sim_utils.py]: a cleaner syntax is to simply type `default_structure` as the argument of the function, it's defined in `conftest.py` and is the consistent way of parameterising tests. Then you just replace "`struct`" with "`default_structure`" ~ you couldn't have guessed this though as we clearly have both the "right" and the "wrong" way in the code in places

**Review by pc494 — COMMENTED** [review]
Just a couple of things I spotted while getting back up to speed.

**Comment by pc494** [comment]
I'm going to review this over the long weekend*, @isabelwood100 is there anything important that I need to know?

*sorry to have snuck in just under the 5pm Friday buzzer (at least in BST)

**Comment by isabelwood100** [comment]
> I'm going to review this over the long weekend*, @isabelwood100 is there anything important that I need to know?
> 
> *sorry to have snuck in just under the 5pm Friday buzzer (at least in BST)

No worries, I don't think there is anything important to know, i'll be around over the weekend so just let me know if theres any issues!

File: diffsims/sims/diffraction_simulation.py [hunk 4]
```diff
@@ -213,3 +221,33 @@ def get_plot(self, g_max, annotate_peaks=True, with_labels=True, fontsize=12):
                 ax.set_ylabel("Intensities (scaled)")
 
         return plt
+
+    def get_plot_x(self, g_max, annotate_peaks=True, with_labels=True, fontsize=12):
```
Comment (pc494) [inline diffsims/sims/diffraction_simulation.py]: is this a duplication of the above?

File: diffsims/sims/diffraction_simulation.py [hunk 5]
```diff
@@ -174,6 +177,8 @@ class ProfileSimulation:
         Magnitudes of scattering vectors.
     intensities : array-like, shape [n_peaks, 1]
         The kinematic intensity of the diffraction peaks.
+
+    ##change this to list qq
```
Comment (pc494) [inline diffsims/sims/diffraction_simulation.py]: This is tbc?

File: diffsims/tests/test_utils/test_sim_utils.py [hunk 6]
```diff
@@ -32,16 +32,10 @@
     simulate_kinematic_scattering,
     is_lattice_hexagonal,
     uvtw_to_uvw,
-    get_holz_angle,
```
Comment (pc494) [inline diffsims/tests/test_utils/test_sim_utils.py]: did we not add these in recently? I feel like they should not be being deleted?

File: diffsims/tests/test_utils/test_sim_utils.py [hunk 7]
```diff
@@ -146,165 +140,30 @@ def test_uvtw_to_uvw(uvtw, uvw):
     np.testing.assert_almost_equal(val, uvw)
 
 
-class TestHolzCalibration:
```
Comment (pc494) [inline diffsims/tests/test_utils/test_sim_utils.py:149]: similar to above?

File: diffsims/tests/test_utils/test_sim_utils.py [hunk 8]
```diff
@@ -146,165 +140,30 @@ def test_uvtw_to_uvw(uvtw, uvw):
...
-        )
-        assert len(scattering_angle) == 2
-        sa_known = np.array([9.84e-3, 6.56e-3])
-        assert np.allclose(sa_known, scattering_angle, rtol=0.001)
+def test_get_intensities_params(default_structure):
+    latt = default_structure.lattice
+    reciprocal_lattice = latt.reciprocal()
+    reciprocal_radius = 0.2
+    unique_hkls, multiplicites, g_hkls = get_intensities_params(reciprocal_lattice, reciprocal_radius)
+    print(g_hkls)
```
Comment (pc494) [inline diffsims/tests/test_utils/test_sim_utils.py]: drop this line, it's just noise.

File: diffsims/generators/diffraction_generator.py [hunk 9]
```diff
@@ -127,7 +128,7 @@ def calculate_ed_data(
         # Obtain crystallographic reciprocal lattice points within `max_r` and
         # g-vector magnitudes for intensity calculations.
         recip_latt = latt.reciprocal()
-        spot_indices, cartesian_coordinates, spot_distances = get_points_in_sphere(
+        spot_indices, spot_coords, spot_distances = get_points_in_sphere(
```
Comment (pc494) [inline diffsims/generators/diffraction_generator.py]: could we call this `cartesian_coordinates` please.

**Review by pc494 — CHANGES_REQUESTED** [review]
In general this looks like it's going in the correct direction. Comments are as follows:
1) You need to merge the master branch in to this so that we can keep the pixstem functionality
2) I think you've duplicate a plotting function in `ProfileSimulation`
3) A couple of minor things that I've marked in this review need tidying up 

Then the big one is, I think we should combine the two prefactors into one, so `get_kinematical_intensities` will have an argument that is "prefactor" which the final line of the function will use:
`peak_intensities = prefactor * (f_hkls * f_hkls.conjugate()).real`

This then cleans up the syntax calling it in a couple of other places, you won't need to set `excitation_error` to `None` for example. You will need to keep track of your arrays a little carefully though.

Think that's a sensible amount to be getting on with in the short term, let me know if anything is unclear.

**Comment by isabelwood100** [comment]
> In general this looks like it's going in the correct direction. Comments are as follows:
> 
> 1. You need to merge the master branch in to this so that we can keep the pixstem functionality
> 2. I think you've duplicate a plotting function in `ProfileSimulation`
> 3. A couple of minor things that I've marked in this review need tidying up
> 
> Then the big one is, I think we should combine the two prefactors into one, so `get_kinematical_intensities` will have an argument that is "prefactor" which the final line of the function will use:
> `peak_intensities = prefactor * (f_hkls * f_hkls.conjugate()).real`
> 
> This then cleans up the syntax calling it in a couple of other places, you won't need to set `excitation_error` to `None` for example. You will need to keep track of your arrays a little carefully though.
> 
> Think that's a sensible amount to be getting on with in the short term, let me know if anything is unclear.

Thank you for looking over it, i'll get onto those changes today :) I'm not sure why those tests got deleted I believe it happened when I had a merge conflict which when i tried to fix went slightly wrong but ill add them back in.

**Comment by pc494** [comment]
This looks like it might be finished now, is that right?

**Comment by isabelwood100** [comment]
> This looks like it might be finished now, is that right?



> This looks like it might be finished now, is that right?

Yes, i've now made all those changes.

**Comment by dnjohnstone** [comment]
Well done @isabelwood100 a very good contribution!

**Pull request merged**